src/tools/illink/src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -372,15 +374,103 @@ internal void MarkEntireType (TypeDefinition type, in DependencyInfo reason)
 			}
 		}
 
+		DependencyAnalyzer<NoLogStrategy<MarkStepNodeFactory>, MarkStepNodeFactory> _analyzer;
 		void Process ()
 		{
-			while (ProcessPrimaryQueue () ||
+			_analyzer.ComputeDependencyRoutine += (List<DependencyNodeCore<MarkStepNodeFactory>> nodes) => {
+				foreach (DependencyNodeCore<MarkStepNodeFactory> node in nodes) {
+					if (node is ProcessCallbackDependencyNode processNode)
+						processNode.Process ();
+				}
+			};
+			_analyzer.AddRoot (new ProcessCallbackDependencyNode (ProcessAllPendingItems), "start");
+			_analyzer.ComputeMarkedNodes ();
+
+			ProcessPendingTypeChecks ();
+
+			bool ProcessAllPendingItems ()
+				=> ProcessPrimaryQueue () ||
 				ProcessMarkedPending () ||
 				ProcessLazyAttributes () ||
 				ProcessLateMarkedAttributes () ||
-				MarkFullyPreservedAssemblies ()) ;
+				MarkFullyPreservedAssemblies ();
+		}
 
-			ProcessPendingTypeChecks ();
+		public class MarkStepNodeFactory (MarkStep markStep)
+		{
+			public MarkStep MarkStep { get; } = markStep;
+		}
+
+		sealed class ProcessCallbackDependencyNode : DependencyNodeCore<MarkStepNodeFactory>
+		{
+			Func<bool> _processAction;
+			DependencyList? _dependencies;
+
+			public ProcessCallbackDependencyNode (Func<bool> action) => _processAction = action;
+
+			public void Process ()
+			{
+				_dependencies = new DependencyList ();
+				if (_processAction ()) {
+					_dependencies.Add (new ProcessCallbackDependencyNode (_processAction), "Some processing was done, continuation required");
+				}
+			}
+
+			public override bool InterestingForDynamicDependencyAnalysis => false;
+
+			public override bool HasDynamicDependencies => false;
+
+			public override bool HasConditionalStaticDependencies => false;
+
+			public override bool StaticDependenciesAreComputed => _dependencies != null;
+
+			public override IEnumerable<DependencyListEntry>? GetStaticDependencies (MarkStepNodeFactory context) => _dependencies;
+
+			public override IEnumerable<CombinedDependencyListEntry>? GetConditionalStaticDependencies (MarkStepNodeFactory context) => null;
+			public override IEnumerable<CombinedDependencyListEntry>? SearchDynamicDependencies (List<DependencyNodeCore<MarkStepNodeFactory>> markedNodes, int firstNode, MarkStepNodeFactory context) => null;
+			protected override string GetName (MarkStepNodeFactory context) => "Process";
+		}
+
+		public class TypeDependencyNode : DependencyNodeCore<MarkStepNodeFactory>
```

Comment:
Future: I think we should differentiate TypeReferenceNode and TypeDefinitionNode (and actually think if we want to model TypeReferenceNode in the first place really). The difference is that there are potentially many more type references than type definitions. And most of the processing we do is on the definition. Type references are generic instantiations, while type definitions are generic definitions - that's one big difference.